scylla-cdc-replicator/src/replication_tests.rs

```diff
@@ -562,4 +664,70 @@ mod tests {
             .await
             .unwrap();
     }
+
+    #[tokio::test]
+    #[should_panic]
+    async fn test_checking_timestamps() {
+        let schema = TestTableSchema {
+            name: "COMPARE_TIME".to_string(),
+            partition_key: vec![("pk", "int")],
+            clustering_key: vec![("ck", "int")],
+            other_columns: vec![("v1", "int"), ("v2", "boolean")],
+        };
+
+        let operations = vec![
+            (
+                "INSERT INTO COMPARE_TIME (pk, ck, v1, v2) VALUES (?, ?, ?, ?)",
+                vec![Int(1), Int(2), Int(3), Boolean(true)],
+            ),
+            (
+                "UPDATE COMPARE_TIME SET v2 = ? WHERE pk = ? AND ck = ?",
+                vec![Boolean(false), Int(1), Int(2)],
+            ),
+        ];
+
+        let session = Arc::new(
+            SessionBuilder::new()
+                .known_node(&get_uri())
+                .build()
+                .await
+                .unwrap(),
+        );
+        setup_tables(&session, &schema).await.unwrap();
+        execute_queries(&session, operations).await.unwrap();
+
+        replicate(&session, &schema.name).await.unwrap();
+
+        compare_changes(&session, &schema.name).await.unwrap();
+        // We update timestamps for v2 column in src.
+        session
+            .query(
+                "UPDATE COMPARE_TIME SET v2 = ? WHERE pk = ? AND ck = ?",
+                (false, 1, 2),
+            )
+            .await
+            .unwrap();
+        compare_timestamps(&session, &schema).await.unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_skipping_non_readable_timestamps() {
+        let schema = TestTableSchema {
+            name: "COMPARE_TIME".to_string(),
+            partition_key: vec![("pk", "int")],
+            clustering_key: vec![("ck", "int")],
+            other_columns: vec![
+                ("v1", "tuple<int, int>"),
+                ("v2", "map<int, int>"),
+                ("v3", "set<frozen<map<int, int>>>"),
+                ("v4", "tuple<map<int, int>>"),
+                ("v5", "frozen<set<frozen<set<frozen<set<int>>>>>>"),
+                ("v6", "tuple<tuple<map<int, int>, set<int>>, int>"),
+                ("v7", "set<frozen<tuple<set<int>, map<int, int>>>>"),
+            ],
+        };
+
+        // Reading timestamps is dependent only on table's schema. This is why we are testing empty columns.
```

Comment:
I don't understand this comment. In order for `compare_timestamps` to do any work, there must be a row present with live cells. Right now there are no operations so there is nothing to do.

I think you should add some updates so that the table has live cells in the end in each column. Then `compare_timestamps` will compare the timestamps of cells in frozen columns and skip the timestamps in non-frozen columns.

Reply:
My intention for this particular test was to test whether our usage of the CQL `WRITETIME` function would result in an error for more complicated types. You are right that no comparison takes place in this test, but the important bit for this test—getting timestamps, does.

In short, comparing timestamps works like this:
```
for every non-pk/ck column:
    if can read timestamp(column):
       original, replicated := get timestamps(column)
       compare timestamps(original, replicated)
 ```
To get timestamps, I use `SELECT WRITETIME(column_name) FROM ...`. Using `WRITETIME` function on non-frozen collections doesn't work and returns an error: `Error message: Cannot use selection function writeTime on non-frozen collections` **no matter if the column in question has any rows**, so even on an empty column it would result in this error. This is the reason I have written this comment.

Reply:
Ok, makes sense.
On the other hand, we still want to test if the function behaves correctly when/if there are actual cells present.

And my point about frozen columns still stands - we still want to check the correctness of timestamp comparison for `tuple` and `frozen<...>` types, and we need live cells for that.

If you want to have a test with no data that's fine, but then please make a separate test that does have live data **and** complex types (both frozen and non-frozen).

Reply:
I have added insertion for every column and a single element-wise update and tuple delete to this test and renamed it to `test_compare_time_for_complicated_types`. I think that should be enough.